src/bp/core/services/middleware/middleware.ts

```diff
@@ -11,30 +13,38 @@ const defaultOptions = {
 }
 
 export class MiddlewareChain {
-  private stack: sdk.IO.MiddlewareHandler[] = []
+  private stack: { mw: sdk.IO.MiddlewareHandler; name: string }[] = []
 
   constructor(private options: MiddlewareChainOptions = defaultOptions) {
     this.options = { ...defaultOptions, ...options }
   }
 
-  use(fn: sdk.IO.MiddlewareHandler) {
-    this.stack.push(fn)
+  use({ handler, name }: sdk.IO.MiddlewareDefinition) {
+    this.stack.push({ mw: handler, name })
   }
 
   async run(event: sdk.IO.Event) {
-    for (const mw of this.stack) {
+    for (const { mw, name } of this.stack) {
       let timedOut = false
       const timePromise = new Promise(() => {}).timeout(this.options.timeoutInMs).catch(() => {
         timedOut = true
       })
-      const mwPromise = Promise.fromCallback<boolean>(cb => mw(event, cb))
-      const result = await Promise.race([timePromise, mwPromise])
+      const mwPromise = Promise.fromCallback<boolean>(cb => mw(event, cb), { multiArgs: true })
+      const result = await Promise.race<Boolean[]>([timePromise, mwPromise])
+
       if (timedOut) {
+        addStepToEvent(`mw:${name}:timedOut`, event)
         continue
       } else if (typeof result !== 'undefined') {
-        // middleware calling next(null, false) will swallow the event
-        if (!result) {
+        const [swallow, skipped] = result as Boolean[]
+
+        if (swallow) {
+          addStepToEvent(`mw:${name}:swallowed`, event)
```

Comment:
again, I'm now a fan of this hardcoded syntax all around

why not
addStepToEvent = (scope, name, status, event) => {....}

Reply:
Good point, i'll change it